pkg/app/zipper/tldcache.go

```diff
@@ -34,12 +45,64 @@ func (app *App) probeTopLevelDomains(ms *PrometheusMetrics) {
 	}
 }
 
+type tldPrefix struct {
+	prefix  string
+	nsCount int
+}
+
+func sortedByNsCount(prefixes []string) []string {
+	countedPrefixes := make([]tldPrefix, len(prefixes))
+	for i, prefix := range prefixes {
+		nsCount := 0
+		if prefix != "" {
+			nsCount = strings.Count(prefix, ".") + 1
+		}
+		countedPrefixes[i] = tldPrefix{
+			prefix:  prefix,
+			nsCount: nsCount,
+		}
+	}
+	sort.Slice(countedPrefixes, func(i, j int) bool {
+		return countedPrefixes[i].nsCount < countedPrefixes[j].nsCount
+	})
+	sortedPrefixes := make([]string, len(prefixes))
+	for i := range countedPrefixes {
+		sortedPrefixes[i] = countedPrefixes[i].prefix
+	}
+	return sortedPrefixes
+}
+
+// getBackendsForPrefix returns the backends that need to be queried in order to populate TLD cache for the prefix.
+// it reuses already fetched tlds to find out about the info. if no info is there, it returns all the backends.
+func getBackendsForPrefix(prefix string, backends []backend.Backend, tldCache map[string][]*backend.Backend) []*backend.Backend {
+	segments := strings.Split(prefix, ".")
```

Comment:
I'm not sure we need to do this every time. As mentioned in another comment, this could be moved to a separate function to process prefixes config.